Fix broken parsing of invokes (closes #1064)

rupertford opened the pull request.

rupertford commented:
Ready for first review. Probably best from @arporter as it uses the fparser2 tree.

rupertford commented:
Summary of changes: a recent improvement to fparser added some additional constraints to expression matching. The arguments to invoke calls are actually structure constructors. However, in the past fparser has always matched these as array references (as it does not know the difference). With the new constraints, fparser now sometime matches invoke arguments correctly as structure constructors. Currently fparser can only do this when there is a non-integer literal specified. PSyclone had been assuming invoke arguments would be parsed as arrays when extracting kernel-call information. As this is no longer always true PSyclone started failing to match some invokes. This PR modifies PSyclone so that it can read kernel-call information when it is matched as a structure constructor or an array. The changes required are relatively small. The only interesting change is that `self.x` has a separate matching rule in a structure constructor. I also found that the code creating a name from an argument that accessed structures was broken so have fixed that e.g. a(0)%b%c(10) should return a_b_c but did not.

arporter commented on src/psyclone/tests/parse/algorithm_test.py:
```diff
@@ -288,7 +316,7 @@ def test_getkernel_noexpr(parser, content):
-def test_getkernel_argerror(monkeypatch, parser):
+def test_getkernel_argerror(monkeypatch):
     '''Test that the get_kernel function raises an exception if it does
     not recognise the fparser2 parse tree for an argument.
```
pylint complains about useless suppression of 'unused-argument' below.

rupertford replied:
Fixed complaints.

arporter commented on src/psyclone/tests/parse/algorithm_test.py:
```diff
@@ -274,7 +302,7 @@ def test_getkernel_isarg(parser, content):
+def test_getkernel_noexpr(content):
     '''Test that the get_kernel function recognises an expression
     containing a variable and raises an exception (as this is not
     currently supported).
```
pylint complains about useless suppression of unused-argument.

rupertford replied:
Fixed complaints.

arporter commented on src/psyclone/tests/parse/algorithm_test.py:
```diff
@@ -1,7 +1,7 @@
 # BSD 3-Clause License
 #
-# Copyright (c) 2019, Science and Technology Facilities Council.
+# Copyright (c) 2019-2021, Science and Technology Facilities Council.
```
There are various pylint warnings in this file (reimport of ParserFactory, imports from fparser are not grouped, lots of useless suppression of unused-argument).

rupertford replied:
Fixed complaints.

arporter commented on src/psyclone/parse/algorithm.py:
```diff
@@ -256,7 +258,8 @@ def create_invoke_call(self, statement):
                 invoke_label = self.check_invoke_label(argument)
 
-            elif isinstance(argument, (Data_Ref, Part_Ref)):
+            elif isinstance(
```
Please could you add/update `__all__` at the bottom of this file.

rupertford replied:
Added.

arporter reviewed (changes requested):
Good job.
All tests pass, incl. with compilation. Ditto examples.
All docs build OK.
Changes look good. Test file just needs updating a bit to clean-up some pylint warnings.

TeranIvy commented:
I tried testing the branch on LFRic and unfortunately I still get this failure
```"Parse Error: algorithm.py:Parser():create_invoke_call: Expecting argument to be of the form 'name=xxx' or a Kernel call but found 'setval_c(some_field, 1.0_r_def)' in file 'working/lfric_atm/algorithm/dummy_alg_mod.x90'."
```
for this simple infrastructure [algorithm](https://code.metoffice.gov.uk/trac/lfric/browser/LFRic/trunk/infrastructure/source/algorithm/dummy_alg_mod.x90).

arporter commented:
@TeranIvy Did you do a `git submodule update`?

TeranIvy commented:
> 
> 
> @TeranIvy Did you do a `git submodule update`?

Yes, I did `git submodule update --recursive` before fparser installation from this branch.

rupertford commented on src/psyclone/parse/algorithm.py:
```diff
@@ -618,34 +629,44 @@ def get_kernel(parse_tree, alg_filename):
+        :py:class:`fparser.two.Fortran2003.Part_Ref` or \
+        :py:class:`fparser.two.Fortran2003.Proc_Component_Ref`
+
+    :returns: a valid variable name as a string.
```
Updated.

rupertford commented:
> Yes, I did `git submodule update --recursive` before fparser installation from this branch.

Hi @TeranIvy. I'm not sure what is happening. PSyclone now works on this file when run directly on it (at least for me locally) and all the others I've tried in the lfric repo work as well. The error implies you are using the latest version of fparser but not this branch (unless there is a different mistake in the branch of course). Could you try running PSyclone directly on the offending file and if it does not work, send the file to me? `psyclone dummy_alg_mod.x90` should be all that is needed. Sorry for the hassle.

rupertford commented:
Hopefully addressed reviewer's comments. Ready for next review. However, this should not be put on master until the issue identified by @TeranIvy is resolved.

TeranIvy commented:
Right, it turns out `git submodule update --recursive` wasn't enough when checking out a working copy of a PSyclone branch with the associated fparser submodule. I had to run `git submodule update --init --recursive` to get the correct commit of the fparser module. This means I'll have to modify my scripts to install PSyclone from source.

I am happy to report that everthing works now. I tested PSyclone on an LFRic branch specifically created to trigger the fparser long comment block bug. The code now builds and runs successfully.

arporter reviewed (approved):
All requested changes have been made.
Tests still all OK, as are examples.
Ref, Dev and User Guides all build OK.
